pyfr/plugins/turbulence.py

```diff
@@ -0,0 +1,270 @@
+
+                    if np.any(candidate):
+                        vinside = vbox.pts_in_region(ptsri[:,candidate,:])
+                        if np.any(vinside):
+                            elestemp = np.any(vinside, axis=0).nonzero()[0].tolist()
+ 
+                    for leid in elestemp:
+                        exmin = ptsri[vinside[:,leid],candidate[leid],0].min()
+                        exmax = ptsri[vinside[:,leid],candidate[leid],0].max()
+                        ts = max(vort['tinit'], vort['tinit'] + ((exmin - xmin - ls)/ubar))
+                        te = max(ts,min(ts + (exmax-exmin+2*ls)/ubar,vort['tinit']+((xmax-xmin)/ubar)))
+                        stream[eids[candidate[leid]]].append((vid,ts,te))
+ 
+                for k, v in stream.items():
+                    v.sort(key=lambda x: x[1]) 
+                    sstream[k] = np.asarray(v, self.sstreamdtype)
+
+                nvmx = 0
+                for leid, actl in sstream.items():
+                    for i, ts in enumerate(actl['ts']):
+                        cnt = 0
+                        while i-cnt >= 0:
+                            if actl['te'][i-cnt] < ts:
+                                break
+                            cnt += 1
+                        if cnt > nvmx:
+                            nvmx = cnt
+                nvmx += 1
+
+                buff = np.zeros((nvmx, neles), self.buffdtype)
+
+                actbuff = {'trcl': 0.0, 'sstream': sstream, 'nvmx': nvmx, 'buff': buff,
+                           'tinit': eles._be.matrix((nvmx, 1, neles), tags={'align'}),
+                           'state': eles._be.matrix((nvmx, 1, neles), tags={'align'}, dtype=np.uint32)}
+
+                eles.add_src_macro('pyfr.plugins.kernels.turbulence','turbulence',
+                {'nvmax': nvmx, 'ls': ls, 'ubar': ubar, 'srafac': srafac,
+                 'ymin': ymin, 'ymax': ymax, 'zmin': zmin, 'zmax': zmax,
+                 'sigma' : sigma, 'rootrs': rootrs, 'gc': gc, 'rot': rot, 'shift': shift
+                }, ploc=True, soln=True)
+
+                eles._set_external('tinit',
+                                   f'in broadcast-col fpdtype_t[{nvmx}][1]',
+                                   value=actbuff['tinit'])
+                                   
+                eles._set_external('state',
+                                   f'in broadcast-col uint32_t[{nvmx}][1]',
+                                   value=actbuff['state'])
+
+                self.actbuffs.append(actbuff)
+                print(f'Rank = {rank}, etype = {etype}, nvorts = {nvorts}, nvmx = {nvmx}.')
+
+        if not bool(self.actbuffs):
+           self.tnext = math.inf
+                  
+    def __call__(self, intg):
+        
+        tcurr = intg.tcurr
+        if tcurr+self.dtol >= self.tnext:
+            for abid, actbuff in enumerate(self.actbuffs):    
```

Comment:
can any of this be put in a function or simplified? Normally this amount of indenting is a sign that it can be